### ISSUE TITLE
Mistake in RMSE usage causes incorrect evaluation metric values

### ISSUE THREAD

**Issue by AndrejKubanda** [robert-bevan/bglp#1]
Model predictions are evaluated incorrectly, which leads to a falsely lower RMSE than it should be.
Your Keras models are compiled with your custom `RMSE` function, which does not consider batched data because it does not accumulate the squared differences across batches before taking the square root of its mean at the end of the evaluation.

Usage of `tensorflow.keras.metrics.RootMeanSquaredError` fixes this problem and leads to correct values.

Examples of the difference between values from the current RMSE function (left) vs Keras metric (right) at 30min horizon with seed 0.

540 RMSE: 20.923 vs 22.970 
544 RMSE: 16.094 vs 17.953 
552 RMSE: 15.686 vs 17.113 
567 RMSE: 20.082 vs 23.602 
584 RMSE: 19.998 vs 22.569 
596 RMSE: 15.779 vs 17.823

**Comment by robert-bevan** [comment]
You're absolutely right - thanks for reporting this.

**Comment by mariogarjim** [comment]
@AndrejKubanda Could you please provide your code? `tensorflow.keras.metrics.RootMeanSquaredError` doesn't work for me as a loss function.

**Comment by AndrejKubanda** [comment]
@rentton Yes, that's right, loss and metric classes are not interchangeable and work differently. And RMSE loss is not implemented in Keras.
However, the training loss function is not the issue here. But you can use the metric also during the training if you compile the model with it. You can also compile the model with both metrics to see the difference in values when batches play a role.
```
model.compile(
    loss='mse',  # or your RMSE or any other
    optimizer='adam',
    metrics=[RMSE, tf.keras.metrics.RootMeanSquaredError()]
)
```   

The real issue is in the evaluation script `eval_model.py`. You use the following code to compute the RMSE:
```
# evaluate model
rmse = config['data_stdev']*model.evaluate(X_test, Y_test)[1]
```
Here, `model.evaluate` evaluates the model against all compiled metrics, which in your case is `RMSE` function (at index 1), which ignores data batches. 

I would recommend to call `model.predict` first, then inverse-transform the values into the original scale, and then use  [tensorflow.keras.metrics.RootMeanSquaredError](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/RootMeanSquaredError) metric for evaluation against the target values:
```
# Predict
y_pred = model.predict(x)
# TODO: inverse transform y_pred to original scale

# Evaluate
rmse_metric = tf.keras.metrics.RootMeanSquaredError()
rmse_metric.update_state(y_target, y_pred)
rmse = rmse_metric.result().numpy()
# don't forget to call rmse_metric.reset_state() if you reuse the object to evaluate multiple patients
```
Alternatively, if the model is compiled with `tf.keras.metrics.RootMeanSquaredError` as one of its metrics, the following code should suffice.
```
rmse = config['data_stdev']*model.evaluate(x, y, return_dict=True)['root_mean_squared_error']
```

The third option would be not to use `model.evaluate` at all (because it works on batches), but to predict the values, and then you can use your `RMSE` function as now all predictions are in the form of a single batch, and thus the results will be equivalent to the previous solutions.

I hope it helps.
